dashboard/app/models/channel_token.rb

```diff
@@ -17,4 +17,46 @@
 class ChannelToken < ActiveRecord::Base
   belongs_to :user
   belongs_to :level
+
+  def self.create_channel_token(level, user, ip, data = {})
+    # The channel should be associated with the template level, if present.
+    # Otherwise the current level.
+    host_level = level.project_template_level || level
+
+    # If `create` fails because it was beat by a competing request, a second
+    # `find_by` should succeed.
+    retryable on: [Mysql2::Error, ActiveRecord::RecordNotUnique], matching: /Duplicate entry/ do
+      # your own channel
+      ChannelToken.find_or_create_by!(level: host_level, user: user) do |ct|
+        # Get a new channel_id.
+        ct.channel = create_channel ip, data
+      end
+    end
+  end
+
+  def self.find_channel_token(level, user)
+    # The channel should be associated with the template level, if present.
+    # Otherwise the current level.
+    host_level = level.project_template_level || level
+
+    ChannelToken.find_by(level: host_level, user: user)
+  end
+
+  # Create a new channel.
+  # @param [Hash] data Data to store in the channel.
+  # @param [String] src Optional source channel to copy data from, instead of
+  #   using the value from the `data` param.
+  def self.create_channel(ip, data = {}, src = nil)
```

Comment:
Channels can be separate from ChannelTokens (e.g. in `projects_controller.rb`) but I think you have the right idea to put it here.  There's no `channel.rb` model and it's much better here than in a helper.